### PR TITLE
enable dynamic menu on linux/windows

### LINKED ISSUE #33457: Doesn't open updated file correctly
There is a file called "foo.txt." I updated via "git pull," but VS Code still gave me an old version until manually closing the file opened by default.

- VSCode Version: Code 1.15.1 (41abd21afdf7424c89319ee7cb0445cc6f376959, 2017-08-16T18:07:25.676Z)
- OS Version: Windows_NT x64 10.0.15063
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
python|don|0.7.0
mdmath|goe|2.0.3
cpptools|ms-|0.12.3
csharp|ms-|1.12.1
rewrap|stk|1.5.0
vim|vsc|0.9.0

---

Steps to Reproduce:

1.
2.

Reproduces without extensions: Yes/No

### LINKED ISSUE #55347: Dynamic items in menu bar can cause crash
- VSCode Version: 1.26.0-ins
- OS Version: win7x64

```
Version: 1.26.0-insider
Commit: dae69dbf00357ff8a3fe740a1f211d66b0b4681c
Date: 2018-07-27T05:17:55.890Z
Electron: 2.0.5
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
Architecture: x64
```

GIF with exts disabled:
crash - menu bar

1) force reload VS Code (or close and start)
2) after 1-2 seconds - open any menu bar when it's options are visible already
3) randomly move cursor above the options

**_Notice_**: with Disabled Extensions Mode it can be detected not so often, really rarely.
BUT in Mode with extensions enabled - every second-third trying to repro - makes a crash!

PS: btw where I can find logs of VS Code crashes?

### PR DESCRIPTION
**Pull request by sbatten**
workaround to fix #55347

### CODE REVIEW

File: src/vs/platform/menubar/electron-main/menubar.ts [hunk 1]
```diff
@@ -36,6 +36,9 @@ export class Menubar {
 	private closedLastWindow: boolean;
 
 	private menuUpdater: RunOnceScheduler;
+	private menuGC: RunOnceScheduler;
+
+	private oldMenus: Menu[];
```
Comment (bpasero) [inline src/vs/platform/menubar/electron-main/menubar.ts]: @sbatten this needs a comment at least why we are doing it with a reference to the electron issue and a TODO to remove this when Electron fixed it!

File: src/vs/platform/menubar/electron-main/menubar.ts [hunk 2]
```diff
@@ -56,6 +59,8 @@ export class Menubar {
 	) {
 		this.menuUpdater = new RunOnceScheduler(() => this.doUpdateMenu(), 0);
 
+		this.menuGC = new RunOnceScheduler(() => { this.oldMenus = this.oldMenus.slice(this.oldMenus.length - 1); }, 10000);
```
Comment (bpasero) [inline src/vs/platform/menubar/electron-main/menubar.ts]: @sbatten I am not sure it helps to always keep the last menu around, I would probably empty the entire array after 10 seconds. Reason is that the user might be lucky to open the menu very early, then we got 2-3 updates and so the last menu in this array will not be the one the user had opened anyways.

File: src/vs/platform/menubar/electron-main/menubar.ts [hunk 3]
```diff
@@ -203,6 +210,11 @@ export class Menubar {
 
 	private install(): void {
 
+		const oldMenu = Menu.getApplicationMenu();
```
Comment (bpasero) [inline src/vs/platform/menubar/electron-main/menubar.ts]: @sbatten again, comment to explain why

File: src/vs/platform/menubar/electron-main/menubar.ts [hunk 4]
```diff
@@ -293,15 +305,6 @@ export class Menubar {
 			menubar.append(macWindowMenuItem);
 		}
 
-		// Preferences
```
Comment (bpasero) [inline src/vs/platform/menubar/electron-main/menubar.ts]: @sbatten is that change related to this PR?

**Review by bpasero — APPROVED** [review]
Some feedback provided

File: src/vs/platform/menubar/electron-main/menubar.ts (hunk 4 above)
Comment (sbatten) [reply src/vs/platform/menubar/electron-main/menubar.ts]: yea this was leftover from before submenu support was added to the custom menubar and we made it a top-level item.
